### ISSUE TITLE
winapp2ool v1.0x development branch bugs

### ISSUE THREAD

**Issue by MoscaDotTo** [MoscaDotTo/Winapp2#380]
Known bugs for the development branch of winapp2ool:

If you use WinappDebug with the file Winapp3.ini, it now deletes the top-level section "Potentially very long scan time (and also dangerous) entries" and all corresponding entries.

**Comment by MoscaDotTo** [comment]
This issue should be fixed now @APMichael

**Comment by APMichael** [comment]
Winapp3.ini works again, but WinappDebug now crashes with Winapp2.ini:
```
Unbehandelte Ausnahme: System.ArgumentOutOfRangeException: Der Index lag außerhalb des Bereichs. Er darf nicht negativ und kleiner als die Sammlung sein.
Parametername: index
   bei System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)
   bei System.Collections.Generic.List`1.get_Item(Int32 index)
   bei winapp2ool.winapp2file.addToInnerFile(Int32 ind, winapp2entry entry, iniSection section)
   bei winapp2ool.winapp2file..ctor(iniFile file)
   bei winapp2ool.WinappDebug.initDebug()
   bei winapp2ool.WinappDebug.handleUserInput(String input)
   bei winapp2ool.Winapp2ool.initModule(String name, Action showMenu, Action`1 handleInput)
   bei winapp2ool.Winapp2ool.handleUserInput(String input)
   bei winapp2ool.Winapp2ool.initModule(String name, Action showMenu, Action`1 handleInput)
   bei winapp2ool.Winapp2ool.main()
```
Translation:
```
Unhandled Exception: System.argumentOutOfRangeException: The index was outside the range. It must not be negative and smaller than the collection.
```

**Comment by MoscaDotTo** [comment]
off by one error :) should be working now

**Comment by APMichael** [comment]
No crash anymore, but after the line "; End of Thunderbird" are two lines with a single semicolon and all entries from [.NET Framework *] to [Zynga Games *] are repeated after the entry [Zynga Games *].

**Comment by MoscaDotTo** [comment]
pesky off by ones :) I got turned around by the disjoint number of headers vs sections being inserted into headers and transposed the numbers. Should really actually be working now

**Comment by APMichael** [comment]
😃 Confirmed! WinappDebug now works as expected. Many thanks for the new top-level sections and the bug fixing.

**Comment by APMichael** [comment]
I downloaded the latest build and noticed that the detection of a "duplicate section name" (exact match) now works on my system too! 😄

But instead of the error message:
```
Error: Duplicate section name detected: [Blob Storage *]
Line: ##
Duplicates the entry on line: ##
This section will be ignored until it is given a unique name.

Press enter to continue.
``` 
I get the following message (translated):
```
Error: System.ArgumentException: An element with the same key has already been added.
   in System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   in System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   in winapp2ool.iniFile.mkSection(List`1 sectionToBeBuilt, List`1 lineTrackingList)
Please report this error on GitHub
Press Enter to continue
```
However, WinappDebug continues to run without errors afterwards.

**Comment by MoscaDotTo** [comment]
Exception handling code generally works in winapp2ool to intercepting the exceptions "message" field and processing based on that. This was short sighted as this breaks for people not using EN-US as their system language. I made a change to this specific error to hopefully fix this by querying the exception type. Download the latest build again and give it a try

**Comment by APMichael** [comment]
Yes, it is fixed! 👍 However, the line number points to the next entry (line 121), not to the duplicate entry (line 75).
```
Error: Duplicate section name detected: [Blob Storage *]
Line: 121
Duplicates the entry on line: 29
This section will be ignored until it is given a unique name.

Press enter to continue.
```

**Comment by MoscaDotTo** [comment]
This should be resolved now

**Comment by APMichael** [comment]
😃 Confirmed! Thank you very much!

**Comment by MoscaDotTo** [comment]
I added a new feature in the latest build 

Diff now will offer to trim winapp2.ini files downloaded from GitHub. The assumption here being that most people have trimmed files and the Diff isn't useful when it says there's 2000 new entries

**Comment by MoscaDotTo** [comment]
Enabled by default if downloading which is now enabled by default if you're online

**Comment by MoscaDotTo** [comment]
by the way, i've been working on implementing logging into winapp2ool recently. Try running stuff and returning the to main winapp2ool menu and entering `printlog` or `savelog`

**Comment by APMichael** [comment]
Great, the log file is very detailed. Maybe Winapp2ool can output something like "Log file is being created, this may take a while".

I found a little bug, probably cause by https://github.com/MoscaDotTo/Winapp2/commit/7ecab3edba26bcc00a0b88d9eb431576d7711217. If you start Winapp2ool and enter 1, 4 and 15, the following error message appears:
```
Error: System.ArgumentOutOfRangeException: The index was out of range. It must not be negative and smaller than the collection.
Parametername: index
   in System.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)
   in System.Collections.Generic.List`1.get_Item(Int32 index)
   in winapp2ool.advSettings.handleUserInput(String input)
   in winapp2ool.Winapp2ool.initModule(String name, Action showMenu, Action`1 handleInput)
```

**Comment by MoscaDotTo** [comment]
Fixed in 2a59a5dc8d98afbebd9b377c1ebd2f0d1dc8f42e

**Comment by APMichael** [comment]
Thanks!

Now I've found a problem with Trim: If you start Winapp2ool and enter 2 (Trim), 2 (Toggle Download) and 1 (Run), Winapp2ool will not download the file and will only output "winapp2.ini was empty or not found". And option 4 (Reset Settings) opens a "File Chooser" instead of resetting the settings.

And a feature request for Trim: The old batch version supported an exclusion list. Is it possible to add this feature? If Trim finds a file named "exclude.ini", all entries listed in that file will be excluded/ignored. 

Example for "exclude.ini":
```
[Cached File Extensions *]
[Cached Shell Extensions *]
[Windows ShellBags *]
```
These 3 entries will not be added to the "winapp2-trimmed.ini".

**Comment by MoscaDotTo** [comment]
You can do that now by merging exclude.ini with winapp2.ini using `Add & Remove` as the merge mode. I don't have tests for this yet so I'm not sure that it works but that is the intended functionality. This should remove any entries from winapp2.ini whose names match the ones in exclude.ini, but I'll look into refining this more as a feature

**Comment by MoscaDotTo** [comment]
The trim bug is fixed in https://github.com/MoscaDotTo/Winapp2/commit/a856dee95e1b3f00d2c2f3a9e33e721068393d95

**Comment by APMichael** [comment]
> 
> 
> You can do that now in a roundabout way by merging exclude.ini with winapp2.ini using `Add & Remove` as the merge mode. I don't have tests for this yet so I'm not sure that it works but that is the intended functionality. This should remove any entries from winapp2.ini whose names match the ones in exclude.ini, but I'll look into refining this more as a feature

Ok, thanks, look like I missed that. It works partely. If I use an "exclude.ini" like in my example, the entries are not completely removed. The names of the entries (which are not detected on the system, but are listed in the "exclude.ini") are still in the target file. But that's the intended function of Merge! Maybe you can add a "Remove only" (without "Add") mode in a future version.

> 
> 
> The trim bug is fixed in [a856dee](https://github.com/MoscaDotTo/Winapp2/commit/a856dee95e1b3f00d2c2f3a9e33e721068393d95)

I'm afraid not. Now option 1 (Run) opens the "File Chooser" instead of downloading the file. And option 4 (Reset Settings) still opens the "File Chooser" instead of resetting the settings.

**Comment by APMichael** [comment]
BTW, Trim detects this entry for Windows 8/8.1 on my Windows 10:
```
[Sharing MFU *]
DetectOS=6.2|6.3
LangSecRef=3025
Detect=HKCU\Software\Microsoft\Windows
Default=False
Warning=This will remove the frequently shared list. This does not remove the apps listed under sharing.
RegKey1=HKCU\Software\Microsoft\Windows\CurrentVersion\Explorer\SharingMFU
```

**Comment by MoscaDotTo** [comment]
Try https://github.com/MoscaDotTo/Winapp2/commit/3f3204a5f7c30797c5e58394e494eb7db05895cb

**Comment by APMichael** [comment]
Trim (Download, Run and DetectOS) now works as expected! 👍

However, if you have enabled the download, option 4 (Reset Settings) still opens the "File Chooser" of option 3 instead of resetting the settings. By the way, you can enter any number from 4 up, it always opens the "File Chooser", the message "Invalid input. Please try again" does not appear.

**Comment by APMichael** [comment]
Problem with Diff: Enter 1 (Run) without a winapp2.ini file and the "File Chooser" will appear. If you then enter 0 (Exit), the following error message appears:
```
Error: System.NullReferenceException: The object reference was not defined for an object instance.
   in winapp2ool.MenuMaker.replDir(String dirStr)
   in winapp2ool.iniFile.printFileChooserMenu()
   in winapp2ool.Winapp2ool.initModule(String name, Action showMenu, Action`1 handleInput)
```
Option 6 (Reset Settings) seems to set the old default values.

**Comment by MoscaDotTo** [comment]
https://github.com/MoscaDotTo/Winapp2/commit/18665f628066050eb010afeb35c8dc79898571f7

Hows this work? You may find NullReferenceException errors in other modules yet as I recently refactored the code that enforced that files have contents (its much better now) but haven't installed all the enforcement plumbing yet

**Comment by APMichael** [comment]
Thank you. Trim now works perfectly!

Diff does not work with download enabled: If you start Winapp2ool and enter 4 (Diff) and 1 (Run), the error message now appears directly. This also happens if you select the older file first.

I also noticed that the status of the "Remote file trimming" option also affects when the download is turned off.

**Comment by MoscaDotTo** [comment]
https://github.com/MoscaDotTo/Winapp2/commit/e8812950008d935a11bb21090ebd693c85e9039e

local files shouldn't be trimmed anymore in this build, I adjusted the validation again too

**Comment by APMichael** [comment]
Diff now works as expected again! 👍

I could cause another NullReferenceException error in the following way:
Start Winapp2ool and enter 4 (Diff), 1 (Run), 0 (Exit), 3 (Online), 4 (File Chooser).
But you wrote in your previous post that this could still happen.

**Comment by MoscaDotTo** [comment]
https://github.com/MoscaDotTo/Winapp2/commit/e8719caa24fa1381ff430f6cecb6612e52de356e

**Comment by APMichael** [comment]
No more NullReferenceException errors. 😃 Thanks a lot.

**Comment by APMichael** [comment]
The most recent build ignores option 5 (Toggle Log Saving) of Diff. The log file (diff.txt) is always written.

**Comment by MoscaDotTo** [comment]
Try https://github.com/MoscaDotTo/Winapp2/commit/bb0ec885a208d6d51bf2a721426fda8f47946335

**Comment by APMichael** [comment]
Works! Thanks! 👍

In WinappDebug I now noticed a long delay before it showed "Press any key to return to the menu.".

**Comment by MoscaDotTo** [comment]
WinappDebug's log can be extremely large with an untrimmed file, for now I've deferred generating it until the Log Viewer option is requested. 

https://github.com/MoscaDotTo/Winapp2/commit/6d87cf48c92025bd8fb92d9f74f78f396ec82ed6

**Comment by APMichael** [comment]
The delay's gone. Ok, thanks for the clarification.

**Comment by MoscaDotTo** [comment]
The Diff output has been drastically changed in the recent builds, how do you find it? I think it's much cleaner with the entries not shown by default

**Comment by CSGalloway** [comment]
What is the link for the dev. branch please?

**Comment by MoscaDotTo** [comment]
https://github.com/MoscaDotTo/Winapp2/tree/Branch1

**Comment by CSGalloway** [comment]
How about a link to download the exe ?

**Comment by MoscaDotTo** [comment]
https://github.com/MoscaDotTo/Winapp2/tree/Branch1/winapp2ool/bin/Debug

**Comment by APMichael** [comment]
> 
> 
> The Diff output has been drastically changed in the recent builds, how do you find it? I think it's much cleaner with the entries not shown by default

Yes, that's right, the output is now much cleaner and easier to read. 👍

I have noticed that the lines "Old" and "New" are interchanged in "Modifed Keys":
```
Beginning diff between version 190617 and version 190617

  Backup * has been modified
    Modifed Keys:
        FileKey1
          Old: FileKey1=%LocalAppData%\Google\Chrome*\Application|old_chrome.exe
          New: FileKey1=%LocalAppData%\Google\Chrome*\Application_OLD_WA2|old_chrome.exe
        FileKey2
          Old: FileKey2=%ProgramFiles%\Google\Chrome*\Application|old_chrome.exe
          New: FileKey2=%ProgramFiles%\Google\Chrome*\Application_OLD_WA2|old_chrome.exe
    
  Added entries: 0
  Modified entries: 1
  Removed entries: 0
Diff complete
```

**Comment by MoscaDotTo** [comment]
https://github.com/MoscaDotTo/Winapp2/commit/e9432c9162c8254817d2732ed65848d2f3e053cd

**Comment by APMichael** [comment]
It's fixed! 😃 Many thanks again for the many improvements and the constant further development of Winapp2ool.

**Comment by CSGalloway** [comment]
Robbie,

1) It would  be great if the winapp2ool would check for updates for
whatever branch it was retrieved from.  the branh1 EXE only checks for the
updates on the master branch, etc....

2) It would be great if there would be an option for deleting the local
winapp3.ini after the following command is finished.....

    C:> "C:\My Programs\WinAppTool\winapp2ool.exe" -3 -2d .\winapp3.ini -s

3) the -s option is still wrting stuff to the screen or doing screen
clears..... try the attached custom.bat and the custom.ini and etc....
  a) start CMD and go to where custom.bat resides
  b) start custom.bat
  c) notice the echo statements in the file - should be the only thing on
the screen.....

I have included folder names in the attached 7zip.......

Download link:
  https://1drv.ms/u/s!Ap7yFWW-Hrgesk-Jb8VPl3_pbGAp?e=dnZo9D

**Comment by MoscaDotTo** [comment]
I don't see any output from winapp2ool running your custom.bat, could you provide a screenshot? the -s flag is ignored sometimes, eg. when exceptions are thrown.

**Comment by CSGalloway** [comment]
Robbie,

Add a CLS as the first line, then run custom.bat, then send me a screenshot
please....    it should only show my echo statements....

CSGalloway at gmail  .....

**Comment by CSGalloway** [comment]
"C:\My Programs\WinAppTool\winapp2ool.exe" -3 -2d .\winapp3.ini -s
echo statement above causes the screen to clear

**Comment by MoscaDotTo** [comment]
winapp2ool should now notify you when newer versions of the beta are available if you are running a beta build 

fwiw it does this by downloading the executable from GitHub (to your %temp% directory) and querying its assembly information, I didn't test whether or not antivirus vendors find this suspicious or not. 

![image](https://user-images.githubusercontent.com/529203/59928755-6e729700-940d-11e9-888b-58aab8ab03da.png)

**Comment by MoscaDotTo** [comment]
I neglected to mention that the console clearing bug should be fixed too @CSGalloway

**Comment by CSGalloway** [comment]
Thank you. Will try it out sometime this weekend....

**Comment by ROCKNROLLKID** [comment]
There appears to be a number of false positives with WinappDebug. For example, it is telling me that EC XTRA Timer at line 7269 belongs at line 6922, which is in between DriverUpdate Backups and DriverUpdate Downloads.

**Comment by MoscaDotTo** [comment]
I'm not seeing any output from WinappDebug on the latest winapp2.ini

**Comment by ROCKNROLLKID** [comment]
After a redownload, it no longer happens. Maybe it was just some error in the download or in the file or something.

**Comment by CSGalloway** [comment]
1) I did the EXE update through the EXE and these are the file information.

06/21/2019  11:34 AM           159,232 winapp2ool v1.2.7111.18271.exe.bak
06/22/2019  11:57 PM           120,320 winapp2ool.exe

Something tells me it's not updating with the same branch - not with a
difference of 38,912 bytes....

2) it's still doing the clear screen when I run the custom.bat in CMD....

**Comment by MoscaDotTo** [comment]
I can't reproduce the console clearing in 1.2.7111.32051

**Comment by MoscaDotTo** [comment]
![image](https://user-images.githubusercontent.com/529203/59979519-ff38a680-95b6-11e9-822b-87329a022cea.png)

**Comment by MoscaDotTo** [comment]
![image](https://user-images.githubusercontent.com/529203/59979530-2000fc00-95b7-11e9-98d4-0eee633d87be.png)

**Comment by CSGalloway** [comment]
Seems to work for me now. The other thing is to ask does the debug version
of branch 1 check for updates to the debug version of branch 1?

**Comment by MoscaDotTo** [comment]
Yes. Generally speaking the beta builds are all debug builds, release builds are done to make sure that the application still works without a debugger attached and for publishing non-beta releases.

**Comment by CSGalloway** [comment]
So branch 1 only contains debug builds?

And master contains only non-debug builds?

**Comment by MoscaDotTo** [comment]
Both branches have both, as Branch1 is simply the "pending future state" of the master branch. 

If downloading from Master, one should use the Release build. Release builds have debugging information stripped from them as part of their optimization process, which makes them not so useful for Branch1's purpose, so if downloading from Branch1 (ie. any beta build), one should use the Debug build

**Comment by MoscaDotTo** [comment]
There was an error in a recent build that broke update notifications, it's been fixed now but you'll probably have to manually update the tool if you updated to an afflicted version.

**Comment by CSGalloway** [comment]
Might be nice to post the affected build numbers.....

currently using 1.2.7113.24127

**Comment by CSGalloway** [comment]
Error with the current branch 1 debug:

============
Error: System.UnauthorizedAccessException: Access to the path
'C:\Users\Galloway\AppData\Local\Temp\winapp2ool.exe' is denied.

   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.File.InternalDelete(String path, Boolean checkHost)
   at winapp2ool.downloadr.download(iniFile pathHolder, String link,
Boolean prompt, Boolean quietly)
   at winapp2ool.updater.checkUpdates(Boolean cond)
Please report this error on GitHub
Press Enter to continue
============

**Comment by MoscaDotTo** [comment]
1.2.7119.22428 now available should address this

**Comment by APMichael** [comment]
winapp2file.vb: `Dim langSecRefs As New List(Of String) From {"3029", "3005", "3027", "3026", "3030", "Language Files", "Dangerous Long", "Dangerous", "Microsoft Edge Insider"}`

Please replace `"3005"` with `"3006"`, as the winsys.ini of the latest CC version shows that they will use "3006" for the new "Edge Chromium".

**Comment by decofona** [comment]
@APMichael Hi I looked into it, he has not to replace 3005 with 3006, he just needs to add it like:
3006 - Edge Chromium

**Comment by APMichael** [comment]
@decofona The replacement is correct and @MoscaDotTo has already done it. The change was for WinappDebug's detection of the new "Edge Chromium". WinappDebug does not need a detection of the old "Edge" (= 3005).

**Comment by decofona** [comment]
@APMichael Okay you right.

**Comment by MoscaDotTo** [comment]
https://forum.piriform.com/topic/32310-winapp2ini-additions/page/323/?tab=comments#comment-309591

leaving a link to this comment here so it doesn't get lost to the sands of time

**Comment by APMichael** [comment]
I found a bug in WinappDebug.

If the Keys are misspelled (upper/lower case), there are some strange counting errors, e.g:
```
Filekey1= > FileKey11=
Excludekey2= > ExcludeKey22=
Regkey1= > RegKey11=
```
With FileKeys even their contents are deleted.

**Comment by MoscaDotTo** [comment]
This should be resolved now, try it out

**Comment by APMichael** [comment]
Confirmed! Thank you very much!

But now there is another error: If there is no winapp2.ini file in the home directory of Winapp2ool, it crashes because it could not find any winapp2.ini file in the subdirectory "Debug".
```
Error: System.IO.FileNotFoundException: Die Datei "D:\xxx\Winapp2\Debug\winapp2.ini" konnte nicht gefunden werden.
Dateiname: "D:\xxx\Winapp2\Debug\winapp2.ini"
```

**Comment by MoscaDotTo** [comment]
This should be resolved now - when I changed how update checks are performed, I neglected to abort the operation after observing that a local file doesn't exist

**Comment by APMichael** [comment]
Works! 😃 Many thanks for the explanation and the fix.

**Comment by MoscaDotTo** [comment]
winapp2ool's new update feature may be causing false positives on anti virus software, i'll look into changing it to be less... suspicious... if they don't catch on that it's a false positive.

**Comment by APMichael** [comment]
WinappDebug: Checking for the use of "wrong" slashes (/) should be disabled for RegKeys because they can actually be used in Registry keys and values. An example is the new entry [Locate32 \*].

Is a new stable release planned soon? Edge Chromium is out now, but the last stable release v1.02 does not yet support LangSecRef=3006. And with a new release the "Default=" lines could be removed.

**Comment by MoscaDotTo** [comment]
I was delaying the release of an update for two main reasons: incomplete (new) human readable documentation (the READMEs for each vb file) and the false positive antivirus issue. 

I was never able to get to the root of what caused the false positives; but they seem to be best mitigated by avoiding submitting winapp2ool to virustotal / hybrid analysis. My best guess is that by submitting it to these services, false positives get propagated across multiple anti virus vendors. I think the manner in which I read winapp2ool's 

We can reasonably update winapp2ool without the *all* of the new documentation and add it later. 

I'll make the adjustment for ignoring slashes inside of RegKeys.

**Comment by APMichael** [comment]
Okay, thanks for the reply.

Maybe it's also due to behavior detection on VirusTotal? Winapp2ool immediately downloads files into the Temp folder. This is probably considered suspicious behavior. At least the last build is still not flagged by Windows Defender. 😃

Another thing I noticed: Is it possible that the line numbers are not reset after a check? For me, the line numbers increase with every run.

**Comment by MoscaDotTo** [comment]
I've submitted newer copies of winapp2ool and its source to various antivirus vendors for review and hopefully to resolve the issue. Microsoft has removed winapp2ool from their definitions (this was my main concern as Defender is aggressively built into Windows 10 and will get in the way if it thinks winapp2ool is malicious) 

Awaiting responses from Ad-Aware, ALYac, Arcabit, BitDefender, Emsisoft, and eScan.

**Comment by MoscaDotTo** [comment]
False positives appear to have been resolved 

https://www.virustotal.com/gui/file/2f2df4fa5810f1b0ef24ea8156637568f7118211ecb24df902727ef2872e32b6/detection

**Comment by APMichael** [comment]
It seems that if one scanner starts with a false positive, all the others will take it over without checking. 😞 I guess we have to live with it...

**Comment by MoscaDotTo** [comment]
Seems that we are back on someone's lists... Unfortunately beyond re-submitting in-dev builds of winapp2ool to these vendors I do not have an resolution for this.

**Comment by ROCKNROLLKID** [comment]
I see there is even more AVs flagging the development build then there was last week. Have you been able to pinpoint which component or commit has triggered them all?

**Comment by MoscaDotTo** [comment]
Unfortunately I've been quite busy with my day job recently and haven't had time to give this proper attention, I'll be free to do so over the next week or so. 

Here's an updated virustotal scan from my most recent local build. 

https://www.virustotal.com/gui/file/51d98bec5433e91756e58cf95faa48b4d83222f3be81bec43cf5b433ec2b3727/detection

I'll spend some time over the next week working with vendors to figure out which specific aspects of the application are triggering these detections in spite of having reported the false positives and how these false positives can be avoided in the future

**Comment by MoscaDotTo** [comment]
A new beta build is available with a shiny new feature: saving winapp2ool's settings to disk! This also comes with a new Winapp2ool Settings top-level menu (option 7) 

![image](https://user-images.githubusercontent.com/529203/75597159-d760e300-5a61-11ea-851a-1ce171546bbd.png)

![image](https://user-images.githubusercontent.com/529203/75597165-e6479580-5a61-11ea-843d-7170d0f83e3e.png)

![image](https://user-images.githubusercontent.com/529203/75597185-0a0adb80-5a62-11ea-844e-198213ddbe87.png)

**Comment by CSGalloway** [comment]
The link to download the beta might be helpful !

**Comment by MoscaDotTo** [comment]
It should be available as an automatic update from within any existing build of winapp2ool beta

![image](https://user-images.githubusercontent.com/529203/75597635-96b69900-5a64-11ea-8d67-ad2c75aedc24.png)

However you can always download winapp2ool.exe beta manually from here: https://github.com/MoscaDotTo/Winapp2/blob/Branch1/winapp2ool/bin/Debug/

**Comment by APMichael** [comment]
Many thanks for the new version. 😃

If I use 2 (Trim) > 4 > 2, Winapp2ool crashes:
```
Unhandled exception: System.Collections.Generic.KeyNotFoundException: The specified key was not listed in the dictionary.
   at System.ThrowHelper.ThrowKeyNotFoundException()
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at winapp2ool.settingsHandler.updateSettings(String targetModule, String settingName, String newVal)
   at winapp2ool.SettingsManager.changeFileParams(iniFile& someFile, Boolean& settingsChangedSetting, String callingModule, String settingName)
   at winapp2ool.Trim.handleUserInput(String input)
   at winapp2ool.MenuMaker.initModule(String name, Action showMenu, Action`1 handleInput, Int32 itmLen)
   at winapp2ool.Winapp2ool.handleUserInput(String input)
   at winapp2ool.MenuMaker.initModule(String name, Action showMenu, Action`1 handleInput, Int32 itmLen)
   at winapp2ool.Winapp2ool.main()
```
And just to make sure that the following issues are not forgotten:

1)
> https://forum.piriform.com/topic/32310-winapp2ini-additions/page/323/?tab=comments#comment-309591
> leaving a link to this comment here so it doesn't get lost to the sands of time
2)
> WinappDebug: Checking for the use of "wrong" slashes (/) should be disabled for RegKeys because they can actually be used in Registry keys and values. An example is the new entry [Locate32 *].
3)
> Another thing I noticed: Is it possible that the line numbers are not reset after a check? For me, the line numbers increase with every run.

**Comment by MoscaDotTo** [comment]
A new beta build is available that should fix crashing when changing settings, not being able to find 32bit applications with the word "software" in their registry path, line numbers in winappdebug not resetting on consecutive runs, and disables checks for forward slashes in RegKeys. It should show up as an available update when you open winapp2ool beta.

**Comment by MoscaDotTo** [comment]
by the way! the settings are loaded from disk before commandline args are processed which means that one can feed fewer parameters into the tool if they're not using the defaults by saving them into winapp2ool's settings

**Comment by APMichael** [comment]
Great! Thanks for the quick fix. 😃

**Comment by MoscaDotTo** [comment]
I am awaiting responses from several anti virus vendors regarding winapp2ool's false positive status and insight into what I can change about winapp2ool so as to reduce the frequency of them

**Comment by MoscaDotTo** [comment]
https://www.virustotal.com/gui/file-analysis/OGUwN2E0NzhmMWEwNTEzYWQwZTNiZWFhMTRmNjhhNDM6MTU4MzI0Mjg2NA==/detection

**Comment by APMichael** [comment]
I guess the automatic and immediate download of the .exe file into the temp folder must be disabled to not be classified as a trojan. Maybe it is already enough if the download is executed only after a user prompt?

**Comment by MoscaDotTo** [comment]
I don't know if the executable is the problem. Winapp2.ini is a configuration file and is filled with the names of antivirus applications (this fact is noted by VirusTotal or HybridAnalysis) which is also suspicious. So far none of the vendors that have responded have lent any guidance for how false positives can be avoided in the future (but they're happy to note the non malicious nature of winapp2ool and exclude certain versions from their definitions)

**Comment by MoscaDotTo** [comment]
The problem may be compounded by winapp2ool downloading a version of winapp2ool from GitHub that is itself classified as malicious, however.

**Comment by APMichael** [comment]
Sounds plausible. Maybe we could maintain a plain text file on GitHub with the version numbers and use it for update notifications. The actual download of the files should only take place if the user explicitly selects this from the menu.

It is a real pity that the vendors cannot or do not want to help... 😞

**Comment by MoscaDotTo** [comment]
That's the existing behavior on the Release branch, only the beta downloads the executable to read its version number because the version number is based on the compile time (which means higher numbers are newer). Since releases are much less frequent than beta builds, this is sensible for them and also maintains compatibility with winapp2oolXP which cannot under any circumstances download winapp2ool.exe due to the .NET Framework 4.5 not supporting TLS 1.2 natively

**Comment by APMichael** [comment]
Okay, thanks for clearing that up. 👍

**Comment by dvdbane** [comment]
hi
can detect-debug-environment behavior cause false positive?

**Comment by MoscaDotTo** [comment]
I'm not certain but I don't think so (at least not on its own). Over the past 9 months I've occasionally compiled a release configuration of winapp2ool but it still falsely identified even without detecting the debug environment

**Comment by APMichael** [comment]
The VirusTotal summary of the last build looks very promising. Only one unimportant scanner reports malware. 😃

https://www.virustotal.com/gui/file/b77331bbc537ce0dffb2d485597033240727c694fdb85c6c71e77852ad8a1fae/detection

**Comment by MoscaDotTo** [comment]
I had submitted a sample to SecureAge APEX some time ago but they never got back to me. FWIW only Kaspersky initially identified this version as malicious (alongside SecureAge) which is a step up from previous versions 

Fingers crossed this is the end of our flagging woes!

**Comment by APMichael** [comment]
Did you change anything? Unfortunately, the latest version is flagged by many scanners again.

https://www.virustotal.com/gui/file/a88d0845e2363bcec216a9fa8d391a2c741ef22982d2fadba0e634e6d028a104/detection

The previous version is still almost free of false positives.

**Comment by MoscaDotTo** [comment]
I'm in the process of making some changes I haven't pushed yet on the beta branch to not download winapp2.ini for update checking if it doesn't already exist in the cwd alongside winapp2ool. Likewise, I'm putting winapp2ool automatic update checking behind a top level setting such that the executable won't automatically be downloaded. 

The most suspicious things in hybrid analysis for winapp2ool are that it both has been flagged as malicious and that it downloads a file (itself) that has been flagged as malicious. 

My hope is that by seeding a few versions to virus total and hybrid analysis that don't download the executable automatically, while still keeping it as an option for those who save their settings or use commandline params, we'll be free of this issue. 

Also, hybrid analysis thinks it's suspicious that we download a configuration file who contains antivirus names, so I'm hoping that additionally not downloading winapp2.ini if the file doesn't already exist will help.

**Comment by APMichael** [comment]
Thank you for the quick reply and explanation.

I also hope that this will reduce the false positives to zero.

**Comment by APMichael** [comment]
What a nasty setback: https://www.virustotal.com/gui/file/d6f194b977c2494177aad1d417be253fdde026d802189f0436da750c94a7b37f/detection

Maybe the online functions should be completely deactivated by default, so that Winapp2ool does not open any connections to the Internet? The user can then activate the online functions manually via the menu if necessary.

**Comment by MoscaDotTo** [comment]
Debug build
https://www.virustotal.com/gui/file/b102a44dba1e24b5479b934b0396a4971a672e4266d1ebab29a112b716becc98/detection

Release build
https://www.virustotal.com/gui/file-analysis/OWRkYzNkZjYyY2EwYWRjNmEwZWQzNzUyNTUyMTBjOTI6MTU4NzQ4MjQ4Mg==/detection

There's no difference between these two builds in terms of *what* they do from the programming perspective. no code changes, just different compilation options. Perhaps the issue lies with debugging artifacts in the code? 

I think the best approach might be to try to resolve our issue with SecureAge APEX if possible and I'll move to building for release whenever the beta updates. I'll make the necessary changes for this and see how it goes

**Comment by APMichael** [comment]
It is interesting to see how differently the two builds are rated by the scanners. The debugging artifacts seem to really have a big part in this.

Unbelievable how much work and trials you have to do here just to get rid of these false positives. The analysis of the scanners does not seem to be very smart even in 2020. And the worst thing is that the false positives are simply adopted unchecked by many other vendors. As great as the idea of VirusTotal was, the false positives for many, even very well-known programs, are now annoying... 😞

**Comment by MoscaDotTo** [comment]
I've updated winapp2ool on the master branch to 1.3.0.0, so I'm going to close this. I think it has long since served its purpose as a meta bug and we can continue false positive discussion on on #700
